Main.hs

```diff
@@ -428,41 +457,181 @@ compileInstRule cs ty = case unSpine  $ ty of
     where dropPi = underAbstr a b (compileInstRule cs . unEl)
   _ -> __IMPOSSIBLE__
 
-compileInstanceClause :: LocalDecls -> Clause -> TCM (Maybe (Hs.InstDecl ()))
+-- Plan:
+--  - ✓ Eta-expand if no copatterns (top-level)
+--  - ✓ drop default implementations and chase definitions of primitive methods in minimal records + *checks*
+--  - ✓ compileInstanceClause on resulting clauses
+--
+-- *checks*
+--  - ✓ Only one minimal record
+--  - ✓ all primitives of the minimal are projected from the same dictionary
+--  - ✓ default implementation that get dropped are also projected from that same dictionary
+
+etaExpandClause :: Clause -> C [Clause]
+etaExpandClause cl@Clause{clauseBody = Nothing} = genericError "Instance definition with absurd pattern!"
+etaExpandClause cl@Clause{namedClausePats = ps, clauseBody = Just t} = do
+  case t of
+    Con c _ _ -> do
+      let fields = conFields c
+      let cls = [ cl{ namedClausePats = ps ++ [unnamed . ProjP ProjSystem <$> f],
+                      clauseBody      = Just $ t `applyE` [Proj ProjSystem $ unArg f] }
+                | f <- fields ]
+      return cls
+    _ ->
+      genericDocError =<< fsep (pwords $ "Type class instances must be defined using copatterns (or top-level" ++
+                                         " records) and cannot be defined using helper functions.")
+
+compileInstanceClause :: LocalDecls -> Clause -> C ([Hs.InstDecl ()], [QName])
 compileInstanceClause ls c = do
   -- abuse compileClause:
   -- 1. drop any patterns before record projection to suppress the instance arg
   -- 2. use record proj. as function name
   -- 3. process remaing patterns as usual
   case dropWhile (isNothing . isProjP) (namedClausePats c) of
-    []     -> genericDocError =<< fsep (pwords $ "Type class instances must be defined using copatterns and " ++
-                                                 "cannot be defined using helper functions or record expressions.")
+    [] ->
+      concatUnzip <$> (mapM (compileInstanceClause ls) =<< etaExpandClause c)
     p : ps -> do
       let c' = c {namedClausePats = ps}
           ProjP _ q = namedArg p
 
       -- We want the actual field name, not the instance-opened projection.
       (q, _, _) <- origProjection q
-
-      let uf = hsName (show (nameConcrete (qnameName q)))
-      (_ , x) <- compileClause ls uf c'
       arg <- fieldArgInfo q
-      if visible arg
-        then return $ Just $ Hs.InsDecl () (Hs.FunBind () [x])
-        else return Nothing
-
-fieldArgInfo :: QName -> TCM ArgInfo
+      let uf = hsName $ show $ nameConcrete $ qnameName q
+
+      let
```

Comment:
maybe pulls these helper functions out to the top level? It's getting a bit long.

Reply:
I was hoping Ulf would come along and point me to existing library functions that do the same thing :P. I will try to make these nicer when I touch the codebase again.